spec: cleanup and finalize spec text for date

ryzokuken opened the pull request:
### TODO

- [x] Properties of `Temporal.Date` instances.
- [x] Constructor
- [x] Properties of Constructor
  - [x] `prototype`
  - [x] `@@species`
  - [x] `from`
  - [x] `compare`
- [ ] Properties of `Temporal.Date.prototype`.
- [ ] Abstract Operations

ptomato commented on spec/date.html:
```diff
@@ -821,7 +821,8 @@ <h1>Properties of Temporal.Date Instances</h1>
+      <li>[[Calendar]], representing the calendar system to project the date into.</li>
     </ul>
-    <p>The value of each of these internal slots is a Number which is an integer.</p>
+    <p>The value of each of these internal slots is a Number which is an integer, except for [[Calendar]], whose value is a %Temporal.Calendar% object.</p>
```
It can be any object, not just a Temporal.Calendar object, if you have a user calendar represented by an object with the right methods.

ryzokuken replied:
Done

ptomato commented:
Do you want to go ahead and open a separate PR for the moves of the large text blocks, so that they don't overwhelm the diffs in future reviews?

ryzokuken commented:
> Do you want to go ahead and open a separate PR for the moves of the large text blocks, so that they don't overwhelm the diffs in future reviews?

Sounds like a good idea, actually.

ryzokuken commented:
Moved the abstract ops to the bottom as suggested for better reviewability, but did it in a new commit instead of dropping the old commit because rebasing is hard. We'll squash everything together once it's done anyway, right? Atleast I'll reduce the commit count. Will move all abstract ops blocks to the top once we've fixed the spec in a single PR so it's easy to review and rebase.

Ms2ger commented on spec/date.html:
```diff
@@ -465,24 +468,25 @@ <h1>Properties of Temporal.Date Instances</h1>
-        1. Let _object_ be ? OrdinaryCreateFromConstructor(_newTarget_, `"%Temporal.Date.prototype%"`, « [[InitializedTemporalDate]], [[ISOYear]], [[ISOMonth]], [[ISODay]] »).
+        1. If ! DateTimeWithinLimits(_isoYear_, _isoMonth_, _isoDay_, 12, 0, 0, 0, 0, 0) is not ~in range~, throw a *RangeError* exception.
+        1. If Type(_calendar_) is not *Undefined*, throw a *RangeError* exception.
+        1. If _object_ is not given, set it to ? OrdinaryCreateFromConstructor(%Temporal.Date%, `"%Temporal.Date.prototype%"`, « [[InitializedTemporalDate]], [[ISOYear]], [[ISOMonth]], [[ISODay]], [[Calendar]] »).
```
Let's keep the _newTarget_ argument, rather than duplicating the `OrdinaryCreateFromConstructor` call.

Ms2ger commented on spec/date.html:
```diff
@@ -465,24 +468,25 @@ <h1>Properties of Temporal.Date Instances</h1>
-        1. If _newTarget_ is not given, set it to %Temporal.Date%.
-        1. Let _object_ be ? OrdinaryCreateFromConstructor(_newTarget_, `"%Temporal.Date.prototype%"`, « [[InitializedTemporalDate]], [[ISOYear]], [[ISOMonth]], [[ISODay]] »).
+        1. If ! DateTimeWithinLimits(_isoYear_, _isoMonth_, _isoDay_, 12, 0, 0, 0, 0, 0) is not ~in range~, throw a *RangeError* exception.
+        1. If Type(_calendar_) is not *Undefined*, throw a *RangeError* exception.
```
```suggestion
        1. If Type(_calendar_) is not Object, throw a *RangeError* exception.
```

ryzokuken replied:
oops

ryzokuken commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -1407,6 +1410,19 @@ export const ES = ObjectAssign({}, ES2019, {
+    ES.RejectDate(isoYear, isoMonth, isoDay);
+    ES.RejectDateRange(isoYear, isoMonth, isoDay);
+    if (!calendar || typeof calendar !== 'object') throw new RangeError('invalid calendar');
+    if (object === undefined) object = new GetIntrinsic('%Temporal.Date%');
```
no, this works.

Ms2ger replied:
It wouldn't be an infinite loop, because the constructor never calls it with an undefined `object`. That also means this is dead code, so I don't know how you verified it works. Let's either remove the undefined case for now, or re-inline all this into the constructor.

ryzokuken commented:
rebased on `main` and squashed.

Ms2ger commented on polyfill/lib/date.mjs:
```diff
@@ -14,27 +14,17 @@ import {
+  constructor(isoYear, isoMonth, isoDay, calendar = GetDefaultCalendar()) {
     isoYear = ES.ToInteger(isoYear);
     isoMonth = ES.ToInteger(isoMonth);
     isoDay = ES.ToInteger(isoDay);
```
```suggestion
    isoDay = ES.ToInteger(isoDay);
    calendar = ES.ToTemporalCalendar(calendar);
```

ryzokuken replied:
Done.

Ms2ger commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -1570,6 +1573,30 @@ export const ES = ObjectAssign({}, ES2019, {
+  CreateTemporalDate: (isoYear, isoMonth, isoDay, calendar, object) => {
+    ES.RejectDate(isoYear, isoMonth, isoDay);
+    ES.RejectDateRange(isoYear, isoMonth, isoDay);
+    if (!calendar || typeof calendar !== 'object') throw new RangeError('invalid calendar');
```
```suggestion
```

(We don't have any "assertion" capability in this file, do we?)

ryzokuken replied:
Done.

Ms2ger commented on polyfill/lib/date.mjs:
```diff
@@ -259,14 +249,7 @@ export class Date {
-    const cal1 = GetSlot(one, CALENDAR).id;
-    const cal2 = GetSlot(two, CALENDAR).id;
-    return ES.ComparisonResult(cal1 < cal2 ? -1 : cal1 > cal2 ? 1 : 0);
+    return ES.CompareTemporalDate(one, two);
```
Spec changed to pass the fields separately, and reuses it for `YearMonth`.

ryzokuken replied:
Handled.

Ms2ger commented on polyfill/lib/ecmascript.mjs:
```diff
@@ -1570,6 +1573,30 @@ export const ES = ObjectAssign({}, ES2019, {
+      const val2 = GetSlot(two, slot);
+      if (val1 !== val2) return ES.ComparisonResult(val1 - val2);
+    }
+    const cal1 = GetSlot(one, CALENDAR).id;
+    const cal2 = GetSlot(two, CALENDAR).id;
```
```suggestion
    const cal1 = ES.CalendarToString(GetSlot(one, CALENDAR));
    const cal2 = ES.CalendarToString(GetSlot(two, CALENDAR));
```

with 
```js
  CalendarToString: (calendar) => {
    let toString = calendar.toString;
    if (toString === undefined) {
      toString = GetIntrinsic('%Temporal.Calendar.prototype.toString%');
    }
    return ES.ToString(ES.Call(toString, calendar));
  },
```

This guarantees that the `<`/`>` operators aren't author-observable.

ryzokuken replied:
Done.

Ms2ger commented on spec/abstractops.html:
```diff
@@ -69,6 +69,16 @@ <h1>IsLeapYear ( _year_ )</h1>
+      The IsValidCalendarName abstract operation verifies that the _calendar_ argument (which must be a String value) represents a valid Unicode Calendar Identifier type, according to BCP47.
+    </p>
+    <p>
+      The abstract operation returns true if _calendar_, converted to upper case as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>, is equal to one of the Unicode Calendar Indentifier types, according to BCP47, converted to uppercase as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>. It returns false otherwise.
```
"true" and "false" in bold?

ryzokuken replied:
Done.

Ms2ger commented on spec/abstractops.html:
```diff
@@ -69,6 +69,16 @@ <h1>IsLeapYear ( _year_ )</h1>
+      The IsValidCalendarName abstract operation verifies that the _calendar_ argument (which must be a String value) represents a valid Unicode Calendar Identifier type, according to BCP47.
+    </p>
+    <p>
+      The abstract operation returns true if _calendar_, converted to upper case as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>, is equal to one of the Unicode Calendar Indentifier types, according to BCP47, converted to uppercase as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>. It returns false otherwise.
```
"Unicode Calendar Indentifier types"

* "Identifier", not "Indentifier"
* should "types" be "values"?

Do we not want to support custom identifiers?

ryzokuken replied:
Done. The route for supporting custom identifiers is to patch `Temporal.Calendar.from`, which uses an alternate parser anyway, right?

Ms2ger commented on spec/date.html:
```diff
@@ -831,6 +838,9 @@ <h1>CompareTemporalDate ( _one_, _two_ )</h1>
         1. If _one_.[[ISOMonth]] &lt; _two_.[[ISOMonth]], return -1.
         1. If _one_.[[ISODay]] &gt; _two_.[[ISODay]], return 1.
         1. If _one_.[[ISODay]] &lt; _two_.[[ISODay]], return -1.
+        1. <mark>TODO: make sure this is correct logic.</mark>
+        1. If _one_.[[Calendar]].id &gt; _two_.[[Calendar]].id, return 1.
+        1. If _one_.[[Calendar]].id &lt; _two_.[[Calendar]].id, return -1.
```
`DateDifference` checks that the [[Calendar]] slots are equal in advance, so let's put this in `Date.prototype.compare` itself:

```diff
+        1. Let _result_ be ! CompareTemporalDate(_one_.[[ISOYear]], _one_.[[ISOMonth]], _one_.[[ISODay]], _two_.[[ISOYear]], _two_.[[ISOMonth]], _two_.[[ISODay]]).
+        1. If _result_ ≠ 0, then
+          1. Return _result_.
+        1. Let _calendarOne_ be ? CalendarToString(_one_.[[Calendar]]).
+        1. Let _calendarTwo_ be ? CalendarToString(_two_.[[Calendar]]).
+        1. Let _r_ be the result of performing Abstract Relational Comparison _calendarOne_ &lt; _calendarTwo_.
+        1. If _r_ is *true*, return -1.
+        1. Let _r_ be the result of performing Abstract Relational Comparison _calendarTwo_ &lt; _calendarOne_.
+        1. If _r_ is *true*, return 1.
+        1. Return +0.
```

ryzokuken replied:
Done

Ms2ger commented on spec/date.html:
```diff
@@ -13,20 +13,22 @@ <h1>The Temporal.Date Constructor</h1>
       <li>is the intrinsic object <dfn>%Temporal.Date%</dfn>.</li>
       <li>creates and initializes a new Temporal.Date object when called as a constructor.</li>
       <li>is designed to be subclassable. It may be used as the value of an `extends` clause of a class definition. Subclass constructors that intend to inherit the specified Temporal.Date behaviour must include a super call to the %Temporal.Date% constructor to create and initialize subclass instances with the necessary internal slots.</li>
+      <li>has a `"length"` property whose value is 4.</li>
```
Let's drop this - it should derive from the definition.

ryzokuken replied:
I did it because of precedent on the other constructors in 262, but I will remove if you feel strongly against it.

Ms2ger replied:
I do feel strongly :)

ryzokuken replied:
Done.

Ms2ger commented on spec/date.html:
```diff
@@ -13,20 +13,22 @@ <h1>The Temporal.Date Constructor</h1>
 
     <emu-clause id="sec-temporal.date">
-      <h1>Temporal.Date ( _isoYear_, _isoMonth_, _isoDay_ )</h1>
+      <h1>Temporal.Date ( _isoYear_, _isoMonth_, _isoDay_, _calendar_ )</h1>
```
Should this be optional, though?

ryzokuken replied:
Yes, fixed.

ryzokuken commented on spec/abstractops.html:
```diff
@@ -367,6 +383,9 @@ <h1>ISO 8601 grammar</h1>
           Date
           Date DateTimeSeparator Time
 
+      CalendarDateTime:
```
Once the spec has been updated completely and `DateTime` has been removed everywhere, I would rename this to `DateTime`.

ryzokuken commented on spec/abstractops.html:
```diff
@@ -433,19 +452,19 @@ <h1>ISO 8601 grammar</h1>
     <h1>ParseISODateTime ( _isoString_ )</h1>
     <p>
-      The ParseISODateTime abstract operation examines any |Time| or |DateTime| string accepted by the grammar in <emu-xref href="#sec-temporal-iso8601grammar"></emu-xref> and returns a record with all the date and time components found.
+      The ParseISODateTime abstract operation examines any |Time|, |DateTime| or |CalendarDateTime| string accepted by the grammar in <emu-xref href="#sec-temporal-iso8601grammar"></emu-xref> and returns a record with all the date and time components found.
```
For now, both `DateTime` and `CalendarDateTime` coexist, but I'll fix this eventually moving forwards as mentioned in https://github.com/tc39/proposal-temporal/pull/806#discussion_r497441910.

ptomato commented on spec/date.html:
```diff
@@ -13,20 +13,22 @@ <h1>The Temporal.Date Constructor</h1>
       <emu-alg>
         1. If NewTarget is *undefined*, then
           1. Throw a *TypeError* exception.
+        1. If _calendar_ is *undefined*, set _calendar_ to ? GetDefaultCalendar().
```
We should call this GetISO8601Calendar or GetConstructorDefaultCalendar now, probably in the polyfill code as well.

ryzokuken replied:
Let me know which name is more preferable to you, I don't have any opinions here. I'll make that change in both place.s

ptomato replied:
I wouldn't consider it a blocker for this PR, by the way, so if you want to defer it to a follow up that's fine with me. I think GetISO8601Calendar is most straightforward.

ryzokuken replied:
#984

ptomato commented on spec/date.html:
```diff
@@ -430,9 +443,14 @@ <h1>Temporal.Date.prototype.difference ( _other_ [ , _options_ ] )</h1>
+        1. Let _otherCalendar_ be _other_.[[Calendar]].
+        1. Let _otherCalendarID_ be ? CalendarToString(_otherCalendar_).
+        1. If _calendarID_ ≠ _otherCalendarID_, throw a *RangeError* exception.
+        <!-- 1. Let _largestUnit_ be ? ToLargestTemporalUnit(_options_, « *"hours"*, *"minutes"*, *"seconds"*, *"milliseconds"*, *"microseconds"*, *"nanoseconds"* », *"days"*). -->
```
I think this step should still be present, since the calendar may not do any validation on the options.

ryzokuken replied:
I was thinking that we could just pass in the options bag to the calendar directly and this would allow the function to take extra options based on the calendar.

ryzokuken replied:
(this is also how it works in the polyfill RN, the ISO calendar handles the options inside)

ptomato replied:
OK, that makes sense.

ptomato commented on spec/date.html:
```diff
@@ -430,9 +443,14 @@ <h1>Temporal.Date.prototype.difference ( _other_ [ , _options_ ] )</h1>
+        1. If _calendarID_ ≠ _otherCalendarID_, throw a *RangeError* exception.
+        <!-- 1. Let _largestUnit_ be ? ToLargestTemporalUnit(_options_, « *"hours"*, *"minutes"*, *"seconds"*, *"milliseconds"*, *"microseconds"*, *"nanoseconds"* », *"days"*). -->
+        1. Return ? CalendarDateDifference(_calendar_, _other_.[[ISOYear]], _other_.[[ISOMonth]], _other_.[[ISODay]], _temporalDate_.[[ISOYear]], _temporalDate_.[[ISOMonth]], _temporalDate_.[[ISODay]], _options_).
+        <!-- 1. Return ? CreateTemporalDuration(_result_.[[Years]], _result_.[[Months]], _result_.[[Weeks]], _result_.[[Days]], 0, 0, 0, 0, 0, 0). -->
```
Should this line just be deleted?

ryzokuken replied:
It should be moved once I specify the ISO8601 calendar.

ptomato commented on spec/date.html:
```diff
@@ -538,17 +556,17 @@ <h1>Properties of Temporal.Date Instances</h1>
-        1. If ! DateTimeWithinLimits(_isoYear_, _isoMonth_, _isoDay_, 12, 0, 0, 0, 0, 0) is not ~in range~, then
-          1. Throw a *RangeError* exception.
+        1. If ! DateTimeWithinLimits(_isoYear_, _isoMonth_, _isoDay_, 12, 0, 0, 0, 0, 0) is not ~in range~, throw a *RangeError* exception.
+        1. If Type(_calendar_) is not Object, throw a *RangeError* exception.
```
In the polyfill, it's allowed to pass a string for the calendar, which is run through Temporal.Calendar.from to get a calendar.

ryzokuken replied:
I'll check the polyfill out and change this to match.

ryzokuken replied:
The constructor behaviors match in the polyfill and spec. This is a lower level function than the constructor which doesn't (cannot) exist in the polyfill and I'd prefer we keep things this way.

ptomato reviewed (commented):
I have only a few minor comments aside from the ones from Ms2ger that are still open.

Ms2ger commented on spec/date.html:
```diff
@@ -13,20 +13,24 @@ <h1>The Temporal.Date Constructor</h1>
+        1. If _calendar_ is *undefined*, then
+          1. Set _calendar_ to ? GetDefaultCalendar().
+        1. Else,
+          1. Set _calendar_ to ToTemporalCalendar(_calendar_).
```
```suggestion
          1. Set _calendar_ to ? ToTemporalCalendar(_calendar_).
```

Ms2ger commented on spec/date.html:
```diff
@@ -35,11 +39,11 @@ <h1>Temporal.Date ( _isoYear_, _isoMonth_, _isoDay_ )</h1>
     </ul>
 
-    <emu-clause id="sec-temporal-date-prototype">
+    <emu-clause id="sec-temporal.date-prototype">
```
Please revert

ryzokuken replied:
Done.

Ms2ger commented on spec/date.html:
```diff
@@ -65,16 +69,19 @@ <h1>Temporal.Date.from ( _item_ [ , _options_ ] )</h1>
         The following steps are taken:
       </p>
       <emu-alg>
-        1. Set _options_ to ? NormalizeOptionsObject(_options_).
```
Restore

ryzokuken replied:
Done.

Ms2ger commented on spec/date.html:
```diff
@@ -65,16 +69,19 @@ <h1>Temporal.Date.from ( _item_ [ , _options_ ] )</h1>
         1. Set _result_ to ? RegulateDate(_result_.[[Year]], _result_.[[Month]], _result_.[[Day]], _overflow_).
-        1. Return ? CreateTemporalDateFromStatic(_constructor_, _result_.[[Year]], _result_.[[Month]], _result_.[[Day]]).
+        1. If _calendar_ is *undefined*, set _calendar_ to ! GetDefaultCalendar().
+        1. Set _calendar_ to ? CalendarFrom(_calendar_).
+        1. Set _calendar_ to ? CalendarFrom(_calendar_).
```
I think this is the wrong place; either move it into `ParseTemporalDateString` or into the `Else` block that calls it.

Ms2ger commented on spec/date.html:
```diff
@@ -861,67 +881,6 @@ <h1>BalanceDate ( _year_, _month_, _day_ )</h1>
-      <h1>PadYear ( _y_ )</h1>
-      <emu-alg>
-        1. Assert: _y_ is an integer Number value.
-        1. If _y_ &gt; 999 and _y_ ≤ 9999, then
-          1. Return _y_ formatted as a four-digit decimal number.
-        1. If _y_ ≥ 0, let _yearSign_ be `"+"`; otherwise, let _yearSign_ be `"-"`.
-        1. Let _year_ be abs(_y_), formatted as a six-digit decimal number, padded to the left with zeroes as necessary.
-        1. Return the string-concatenation of _yearSign_ and _year_.
-      </emu-alg>
-    </emu-clause>
-
-    <emu-clause id="sec-temporal-temporaldatetostring" aoid="TemporalDateToString">
-      <h1>TemporalDateToString ( _temporalDate_ )</h1>
-      <emu-alg>
-        1. Assert: Type(_temporalDate_) is Object.
-        1. Assert: _temporalDate_ has an [[InitializedTemporalDate]] internal slot.
-        1. Let _year_ be ! PadYear(_temporalDate_.[[ISOYear]]).
-        1. Let _month_ be _temporalDate_.[[ISOMonth]] formatted as a two-digit decimal number, padded to the left with a zero if necessary.
-        1. Let _day_ be _temporalDate_.[[ISODay]] formatted as a two-digit decimal number, padded to the left with a zero if necessary.
-        1. Return the string-concatenation of _year_, the code unit 0x002D (HYPHEN-MINUS), _month_, the code unit 0x002D (HYPHEN-MINUS), and _day_.
-      </emu-alg>
-    </emu-clause>
-
-    <emu-clause id="sec-temporal-adddate" aoid="AddDate">
-      <h1>AddDate ( _year_, _month_, _day_, _years_, _months_, _weeks_, _days_, _overflow_ )</h1>
-      <emu-alg>
-        1. Assert: _year_, _month_, _day_, _years_, _months_, _weeks_, and _days_ are integer Number values.
-        1. Assert: _overflow_ is either *"constrain"* or *"reject"*.
-        1. Let _y_ be _year_ + _years_.
-        1. Let _m_ be _month_ + _months_.
-        1. Let _intermediate_ be ? BalanceYearMonth(_y_, _m_).
-        1. Let _intermediate_ be ? RegulateDate(_intermediate_.[[Year]], _intermediate_.[[Month]], _day_, _overflow_).
-        1. Set _days_ to _days_ + 7 × _weeks_.
-        1. Let _d_ be _intermediate_.[[Day]] + _days_.
-        1. Let _intermediate_ be ? BalanceDate(_intermediate_.[[Year]], _intermediate_.[[Month]], _d_).
-        1. Return ? RegulateDate(_intermediate_.[[Year]], _intermediate_.[[Month]], _intermediate_.[[Day]], _overflow_).
-      </emu-alg>
-    </emu-clause>
-
-    <emu-clause id="sec-temporal-subtractdate" aoid="SubtractDate">
```
Restore

ryzokuken replied:
Done.

ptomato commented on spec/date.html:
```diff
@@ -767,6 +785,10 @@ <h1>ToPartialDate ( _temporalDateLike_ )</h1>
             </tr>
+            <tr>
+              <td>[[Calendar]]</td>
+              <td>`"year"`</td>
```
```suggestion
              <td>`"calendar"`</td>
```

ptomato reviewed (approved):
This looks ready to merge, with two changes (the one I marked, and I believe there is one comment from Ms2ger still unaddressed).

Ms2ger commented on spec/abstractops.html:
```diff
@@ -100,6 +100,16 @@ <h1>ToTemporalOverflow ( _normalizedOptions_ )</h1>
+      The IsValidCalendarName abstract operation verifies that the _calendar_ argument (which must be a String value) represents a valid Unicode Calendar Identifier value, according to BCP47.
+    </p>
+    <p>
+      The abstract operation returns *true* if _calendar_, converted to upper case as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>, is equal to one of the Unicode Calendar Indentifier types, according to BCP47, converted to uppercase as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>. It returns *false* otherwise.
```
```suggestion
      The abstract operation returns *true* if _calendar_, converted to upper case as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>, is equal to one of the Unicode Calendar Identifier values, according to BCP47, converted to uppercase as described in <emu-xref href="#sec-case-sensitivity-and-case-mapping"></emu-xref>. It returns *false* otherwise.
```
